js-library/src/services/ens.ts

```diff
@@ -0,0 +1,176 @@
+    )
+    this.publicResolverContract = new Contract(
+      PUBLIC_RESOLVER_ADDRESS as string,
+      PublicResolverContract.abi,
+      this._provider,
+    )
+    this.subdomainRegistrarContract = new Contract(
+      SUBDOMAIN_REGISTRAR_ADDRESS as string,
+      SubdomainRegistrarContract.abi,
+      this._provider,
+    )
+  }
+
+  /**
+   * returns RPC provider
+   */
+  public get provider(): providers.JsonRpcProvider {
+    return this._provider
+  }
+
+  /**
+   * Connects signer to the smart contracts
+   * @param signerOrProvider An instance of ethers.js Wallet or any other signer
+   */
+  public connect(signerOrProvider: string | providers.Provider | Signer) {
+    this.publicResolverContract = this.publicResolverContract.connect(signerOrProvider)
+    this.subdomainRegistrarContract = this.subdomainRegistrarContract.connect(signerOrProvider)
+    this.ensRegistryContract = this.ensRegistryContract.connect(signerOrProvider)
+  }
+
+  /**
+   * Returns owner address of the provided username on ENS
+   * @param username ENS username
+   * @returns owner's address
+   */
+  public async getUsernameOwner(username: string): Promise<string> {
+    const usernameHash = this.hashUsername(username)
+
+    const owner = await this.ensRegistryContract.owner(usernameHash)
+
+    return owner
+  }
+
+  /**
+   * Checks whether username is available on ENS
+   * @param username ENS username
+   * @returns True if the username is available
+   */
+  public async isUsernameAvailable(username: string): Promise<boolean> {
+    const owner = await this.getUsernameOwner(username)
+    return owner === NULL_ADDRESS
+  }
+
+  /**
+   * Sets owner of the provided username on ENS
+   * @param username ENS username
+   * @param address Owner address of the username
+   */
+  public async registerUsername(username: string, address: string, publicKey: string): Promise<void> {
+    const ownerAddress = await this.getUsernameOwner(username)
+
+    if (ownerAddress !== NULL_ADDRESS && ownerAddress !== address) {
+      throw new Error(`ENS: Username ${username} is not available`)
+    }
+
+    if (ownerAddress === NULL_ADDRESS) {
+      await waitTransaction(
+        this.subdomainRegistrarContract.register(keccak256(toUtf8Bytes(username)), address),
+      )
+    }
+
+    const usernameHash = this.hashUsername(username)
+
+    await waitTransaction(
+      this.ensRegistryContract.setResolver(usernameHash, this.publicResolverContract.address),
+    )
+
+    await this.setUsernamePublicKey(usernameHash, address, publicKey)
+  }
+
+  /**
+   * Returns public key of registered username
+   * @param username
+   * @returns public key
+   */
+  public async getPublicKey(username: string): Promise<string> {
+    const [publicKeyX, publicKeyY] = await this.publicResolverContract.pubkey(this.hashUsername(username))
+
+    const publicKey = joinPublicKey(publicKeyX, publicKeyY)
+
+    if (isPublicKeyValid(publicKey)) {
+      throw new Error('Public key is not set')
```

Comment:
it is not true, maybe it is just invalid, isn't it?

Reply:
If it contains only zeros then it's not set. But the name of the function is actually incorrect because it doesn't really checks for validity.